### ISSUE TITLE
kitty does not render on mesa 19

### ISSUE THREAD

**Issue by ranisalt** [kovidgoyal/kitty#1484]
Since upgrading to mesa 19, kitty does not render text to screen, making it appear completely blank. It does not render after typing, too. Considering the background is still transparent, it might be an issue with my compositor instead, which I will look into later on.

**Comment by kovidgoyal** [comment]
You can try using --debug-gl to see if any of the OpenGL calls kitty uses return errors, but I doubt that will be the case. Most likely something else is broken.

**Comment by ranisalt** [comment]
I noticed that using a compositor makes kitty unable to render text, but without a compositor the first line is broken and does not show properly. Every other line works as expected, though.

See how the first line renders incomplete:
![image](https://user-images.githubusercontent.com/1993083/54505962-19b51100-4932-11e9-8051-8db6a48dc7c6.png)

**Comment by kovidgoyal** [comment]
I doubt the first line issue has anything to do with compositor/nocompositor. Try running

kitty -o shell=bash

and see if you still get the first line problem.

**Comment by sjclayton** [comment]
@kovidgoyal I can confirm this issue on Arch (on Intel Graphics) as well since upgrading to mesa 19... if I downgrade to the previous mesa version 18.3.4 this issue disappears..

**Comment by kovidgoyal** [comment]
well, I will take a look at it when I have a moment, but really this seems like a bug in mesa 19

**Comment by rgwott** [comment]
@kovidgoyal I can confirm this issue on my gentoo box with kitty 0.13.3. In fact, I had similar issues before when I upgraded mesa from 18.3.4 to 18.3.5, then I downgraded it to keep using kitty. Now I tried mesa 19 and I have the issue with the first line. I use no compositor.

**Comment by buzzlightmonth** [comment]
Can confirm the same issue, though it's not just the first line but all of the lines: some characters render 'backwards' (right to left), and they keep switching around. Almost looks like some sly virus.

Downgrading to 18.x immediately solves the issue.

**Comment by YokaiRick** [comment]
Yep, i had the same "Black Screen" problem (Arch + Kitty and Mesa 19).
Downgrading mesa 18.4.3-1 solved the issue for me too

**Comment by kovidgoyal** [comment]
Well I can reproduce with mesa 19. Since according to @rgwott it broke with mesa 18.3.5 as well, here is the list of changes https://lists.freedesktop.org/archives/mesa-dev/2019-March/216922.html in 18.3.5. Presumably one of them broke something kitty relied on. I dont have the time to go digging in mesa to figure out what exactly broke, so for the moment I suggest simply blicklisting mesa > 18.3.4 to use kitty.

**Comment by kovidgoyal** [comment]
dont see anything obvious in the changelog of mesa, so someone is going to have to find the time to do a bisect on mesa. I have reported it upstream: https://bugs.freedesktop.org/show_bug.cgi?id=110201

**Comment by lferra** [comment]
> Well I can reproduce with mesa 19. Since according to @rgwott it broke with mesa 18.3.5 as well
In my case (Gentoo box) Mesa 18.3.5 seems ok: rendering breaks only with 19

**Comment by alepez** [comment]
I can confirm on Arch with mesa 19. Downgrade to 18.3.4 fix the problem. I'm using compton, but I have the same problem without the compositor. It seems that only the first line is affected.

**Comment by ben-turner** [comment]
I am also able to reproduce the issue and fix it with a downgrade. Something I noticed though is that by mashing Ctrl+C and enter gets me a bunch of mangled characters from my PS1 string.

![image](https://user-images.githubusercontent.com/7623873/54709067-0658a000-4b02-11e9-8eef-f274c1a1d818.png)

Probably not useful information, but I'm adding it just in case.

**Comment by zergera** [comment]
The same (Arch + mesa 19). Downgrade to 18.3.4 (AUR package "downgrade" is a real thing) helped, as suggested.
I would say, It is really creepy! :-)

**Comment by 2ion** [comment]
Just as another variation, in case some users find their way here: Since upgrading to Mesa 19.0.0 there's a black log in the top-left corner of Kitty, always at the same position; lines that go past it get mangled but otherwise the terminals stays usable.

![Screenshot_20190320_214753](https://user-images.githubusercontent.com/666784/54718014-da690900-4b59-11e9-8e49-5470584a5c9b.png)

**Comment by PhotonMike** [comment]
I have done some testing:
* kitty works on the proprietary nvidia driver on nvidia cards, but that's expected, since nvidia doesn't use mesa
* kitty works on AMD cards using the amdgpu driver, and I think that the amdgpu driver uses mesa
* kitty doesn't work on Intel GPUs, at least on the i915 driver

Tested on Arch Linux, with Linux 5.0.3 and Mesa 19.0.0

**Comment by kovidgoyal** [comment]
It is a bug in the intel driver in mesa, see the upstream report I linked to earlier. There is a proposed fix there test it out and post ont he bug if you wish to expedite the fix.

https://bugs.freedesktop.org/show_bug.cgi?id=110201

**Comment by archusr** [comment]
Confirming the patch fixed the issue with kitty for me. Downloaded the [512.diff](https://gitlab.freedesktop.org/mesa/mesa/merge_requests/512.diff
) and added it to the Arch [PKGBUILD](https://www.archlinux.org/packages/extra/x86_64/mesa/) in _prepare_ (similar to a [previous version](https://git.archlinux.org/svntogit/packages.git/commit/trunk?h=packages/mesa&id=d63896f307e6872b8e741a11534c08e874628119)).

**Comment by weskerfoot** [comment]
Downgrading to Mesa 18.3.4 worked for me. If you don't have the package cached you can use [the ABS](https://wiki.archlinux.org/index.php/Arch_Build_System)

For anyone who doesn't know how this works

i.e.
```
svn checkout --depth=empty svn://svn.archlinux.org/packages
svn update mesa
cd packages/mesa/trunk
``
- Edit the line starting with `pkgver` in `PKGBUILD` to be `18.3.4`
- Run `makepkg --skipinteg`, and install any dependencies it complains about
- Then run `sudo pacman -U mesa-18.3.4-1-x86_64.pkg.tar.xz`

It's working fine for me after restarting X11! Thanks!

You can also add `IgnorePkg = mesa` to your `pacman.conf` to avoid upgrading until it's fixed.

**Comment by weskerfoot** [comment]
@archusr So it's fixed in the latest version of the package now?

**Comment by ranisalt** [comment]
@weskerfoot no, that patch was not merged yet.

**Comment by archusr** [comment]
Confirming [520](https://gitlab.freedesktop.org/mesa/mesa/merge_requests/520) fixes the issue with kitty for me.

**Comment by kovidgoyal** [comment]
Cool hopefully they will release it soon

**Comment by ronasimi** [comment]
Mesa 19.01 is in testing, but I didn't see the fix in the commit history.  Can anyone confirm?

**Comment by GladOSkar** [comment]
@ronasimi [520](https://gitlab.freedesktop.org/mesa/mesa/merge_requests/520) hasn't been merged so that makes sense

**Comment by ronasimi** [comment]
OK thanks.

**Comment by archusr** [comment]
@ronasimi You could download my temporary [PKGBUILD](https://gist.github.com/archusr/7594bc0130b513ee694a0d720f6c3958) and run `makepkg -si` in its directory.

**Comment by ronasimi** [comment]
> @ronasimi You could download my temporary [PKGBUILD](https://gist.github.com/archusr/7594bc0130b513ee694a0d720f6c3958) and run `makepkg -si` in its directory.

You're the best, thanks!

**Comment by ronasimi** [comment]
> @ronasimi You could download my temporary [PKGBUILD](https://gist.github.com/archusr/7594bc0130b513ee694a0d720f6c3958) and run `makepkg -si` in its directory.

I'm getting an invalid PGP key error when running makepkg.  I'm gonna be honest, I just stripped the pgp keys out of the pkgbuild, but I wanted to give you a heads up.

**Comment by troubas** [comment]
It seems to happen less often, but I have reproduced using Mesa 18.3.4-1 and 19.0.1.
Running with --debug-gl now to see if I can get some output.

**Comment by ranisalt** [comment]
>  The error is "mesa-19.0.1.tar.xz ... FAILED (unknown public key 4C95FAAB3EB073EC)"

Just run `gpg --recv-keys 4C95FAAB3EB073EC` and it should pass

**Comment by troubas** [comment]
All I'm getting with `--debug-gl` is 
```
GL version string: '4.5 (Core Profile) Mesa 18.3.4' Detected version: 4.5
```
Any way I can get kitty to be more verbose?

Also I noticed I've only downgraded mesa
```
warning: lib32-mesa: local (19.0.1+really+18.3.5-1) is newer than multilib (19.0.1-1)
warning: lib32-mesa-vdpau: local (19.0.1+really+18.3.5-1) is newer than multilib (19.0.1-1)
warning: lib32-vulkan-intel: local (19.0.1+really+18.3.5-1) is newer than multilib (19.0.1-1)
warning: lib32-vulkan-radeon: local (19.0.1+really+18.3.5-1) is newer than multilib (19.0.1-1)
warning: libva-mesa-driver: local (19.0.1+really+18.3.5-1) is newer than extra (19.0.1-2)
warning: mesa: ignoring package upgrade (18.3.4-1 => 19.0.1-2)
warning: mesa-vdpau: local (19.0.1+really+18.3.5-1) is newer than extra (19.0.1-2)
warning: vulkan-intel: local (19.0.1+really+18.3.5-1) is newer than extra (19.0.1-2)
warning: vulkan-radeon: local (19.0.1+really+18.3.5-1) is newer than extra (19.0.1-2)
```
Should I downgrade any of these packages to?

**Comment by kovidgoyal** [comment]
--debug-gl simply causes kiity to print out error reported by OpenGL.
There wont be any here, since this is a bug in a layer well below
opengl, in the intel driver provided by mesa. Essentially they got their
register assignments wrong which causes random number to be read from
textures in some shaders.

**Comment by farid-fari** [comment]
> @ronasimi You could download my temporary [PKGBUILD](https://gist.github.com/archusr/7594bc0130b513ee694a0d720f6c3958) and run `makepkg -si` in its directory.

Great job! Just missing the `flex` dependency.

**Comment by avkoval** [comment]
With 19.0.1-2 I still see some rendering issues, but I am not sure if this is same bug or something else. Exactly, first 3-4 symbols on first line of terminal are not displayed, and sometimes some rendering issues happen on this 1st line, everything else works fine.


![](http://scr.ua2web.com/avk/ss_20190410073452_SseJpt5YrxEWM9vjtxWP.png)

**Comment by kovidgoyal** [comment]
Since the fix was not released as part of mesa 19.0.1 that is hardly surprising. Just downgrade mesa. I will close this bug report when a version of mesa with the fix is released.

**Comment by avkoval** [comment]
Thanks for hint @kovidgoyal I did not realized that. OK, I did it, and it works fine, for other arch users this was the fix:
```
sudo pacman -U /var/cache/pacman/pkg/mesa-18.3.4-1-x86_64.pkg.tar.xz
```

**Comment by ranisalt** [comment]
Actually, prefer not do downgrade and use @archusr PKGBUILD above - Arch has (positive) issues when it comes to library versioning.

**Comment by archusr** [comment]
Temporary PKGBUILD with [520](https://gitlab.freedesktop.org/mesa/mesa/merge_requests/520) updated to 19.0.2
`makepkg -si https://gist.githubusercontent.com/archusr/7594bc0130b513ee694a0d720f6c3958/raw/5900c7ee927245b23ac61e2e885b52e4dfbe9773/PKGBUILD`

**Comment by wallace11** [comment]
I can confirm this bug has been solved with mesa 19.0.2.
I'm running Arch Linux with the latest kitty and mesa 19.0.2-1, both from the official repos.

**Comment by kovidgoyal** [comment]
As far as I know the fix was not included in mesa 19.0.2 so I would be very surprised if the issue was fixed by it.

**Comment by wallace11** [comment]
> As far as I know the fix was not included in mesa 19.0.2 so I would be very surprised if the issue was fixed by it.

I just arrived home and found out that the problem persists on my Ivy Bridge CPU. 
On Kaby Lake it worked just fine earlier today...

Sorry for the false news :(

**Comment by serebit** [comment]
I don't have this issue on my Kaby Lake iGPU on Solus's build of Mesa 19.0.0. Maybe an issue with Arch's build?

**Comment by GladOSkar** [comment]
@serebit Solus with Mesa 19.0.0 on Haswell iGPU does have the problem

**Comment by archusr** [comment]
mesa-19.0.3 available as per https://gitlab.freedesktop.org/mesa/mesa/tags

**Comment by kovidgoyal** [comment]
I cannot reproduce the symptoms on my system with mesa 19.0.3, so closing this bug report.

**Comment by salmanulfarzy** [comment]
`mesa 19.0.2` was published to **Ubuntu 18.04** bionic on [2019-07-03](https://launchpad.net/ubuntu/+source/mesa/19.0.2-1ubuntu1.1~18.04.1) and it broke Kitty for me. Updating to` mesa 19.1.0` with [Padoka stable PPA][1] fixed it.

[1]: https://launchpad.net/~paulo-miguel-dias/+archive/ubuntu/pkppa/?field.series_filter=bionic

**Comment by okdewit** [comment]
Can confirm, on Linux Mint all glitches were fixed using above PPA
```
sudo add-apt-repository ppa:paulo-miguel-dias/pkppa
sudo apt-get update
sudo apt-get upgrade

$ kitty --debug-gl
GL version string: '4.5 (Core Profile) Mesa 19.1.0 - padoka PPA' Detected version: 4.5
```
